src/features/element-hiding.js

```diff
@@ -157,51 +158,83 @@ function isDomNodeEmpty (node) {
     // check at 0ms, 100ms, 200ms, 300ms, 400ms, 500ms, 1000ms, 1500ms, 2000ms, 2500ms, 3000ms
-    setTimeout(function () {
-        hideAdNodes(rules)
-    }, 0)
-
-    let immediateHideIterations = 0
-    const immediateHideInterval = setInterval(function () {
-        immediateHideIterations += 1
-        if (immediateHideIterations === 4) {
-            clearInterval(immediateHideInterval)
-        }
-        hideAdNodes(rules)
-    }, 100)
-
-    let delayedHideIterations = 0
-    const delayedHideInterval = setInterval(function () {
-        delayedHideIterations += 1
-        if (delayedHideIterations === 4) {
-            clearInterval(delayedHideInterval)
-        }
-        hideAdNodes(rules)
-    }, 500)
+    hideTimeouts.forEach((timeout) => {
+        setTimeout(() => {
+            hideAdNodes(timeoutRules)
+        }, timeout)
+    })
 
     // check previously hidden ad elements for contents, unhide if content has loaded after hiding.
     // we do this in order to display non-tracking ads that aren't blocked at the request level
     // check at 750ms, 1500ms, 2250ms, 3000ms
-    let unhideIterations = 0
-    const unhideInterval = setInterval(function () {
-        unhideIterations += 1
-        if (unhideIterations === 3) {
-            clearInterval(unhideInterval)
-        }
-        unhideLoadedAds()
-    }, 750)
+    unhideTimeouts.forEach((timeout) => {
+        setTimeout(() => {
+            unhideLoadedAds(timeoutRules)
+        }, timeout)
+    })
 
     // clear appliedRules and hiddenElements caches once all checks have run
-    setTimeout(function () {
+    setTimeout(() => {
         appliedRules = new Set()
         hiddenElements = new WeakMap()
     }, 3100)
 }
 
+/**
+ * Separate strict hide rules to inject as style tag if enabled
+ * @param {Object[]} rules
+ * @param {string} rules[].selector
+ * @param {string} rules[].type
+ */
+function extractTimeoutRules (rules) {
+    if (!shouldInjectStyleTag) {
+        return rules
+    }
+
+    const strictHideRules = []
+    const timeoutRules = []
+
+    rules.forEach((rule, i) => {
+        if (rule.type === 'hide') {
+            strictHideRules.push(rule)
+        } else {
+            timeoutRules.push(rule)
+        }
+    })
+
+    injectStyleTag(strictHideRules)
+    return timeoutRules
+}
+
+/**
+ * Create styletag for strict hide rules and append it to document head
+ * @param {Object[]} rules
+ * @param {string} rules[].selector
+ * @param {string} rules[].type
+ */
+function injectStyleTag (rules) {
+    let styleTag = '<style>'
+
+    rules.forEach((rule, i) => {
+        if (i === rules.length - 1) {
+            styleTag = styleTag.concat(rule.selector, '{')
+        } else {
+            styleTag = styleTag.concat(rule.selector, ',')
+        }
+    })
+
+    styleTag = styleTag.concat('display:none!important;min-height:0!important;height:0!important;}</style>')
+    document.head.insertAdjacentHTML('beforeend', styleTag)
```

Comment:
This is risking XSS through using a variant of innerHTML; instead we should create the style element, innerText the rules and then append the node.

Reply:
Good catch, updated to address.